storage2/src/state/transaction.rs

```diff
@@ -0,0 +1,43 @@
+use std::collections::HashMap;
+
+use anyhow::Result;
+
+use jmt::storage::{TreeReader, TreeWriter};
+
+use super::{State, StateWrite};
+
+/// Represents a transactional set of changes to a `State` fork,
+/// implemented as a RYW cache over a `State`.
+pub(crate) struct Transaction<'a, R: TreeReader + TreeWriter> {
+    // TODO: should higher-level types be used here instead of jmt types?
+    cache: HashMap<jmt::KeyHash, jmt::OwnedValue>,
+    unwritten_changes: Vec<(jmt::KeyHash, jmt::OwnedValue)>,
+    state: &'a State<'a, R>,
+}
+
+impl<'a, R: TreeReader + TreeWriter> Transaction<'a, R> {
+    pub fn new() -> Self {
+        Self {
+            cache: todo!(),
+            unwritten_changes: todo!(),
+            state: todo!(),
+        }
+    }
```

Comment:
We can't have a `Transaction::new`, because the transaction has to be created with reference to a `State` instance.  We should just have one method on the `State` that produces a `Transaction` (beginning its lifecycle) and a `commit()` method that consumes it (ending its lifecycle).